DOC Update documentation for GraphQL v4 (#10325)
silverstripe/silverstripe-framework: docs/en/02_Developer_Guides/19_GraphQL/01_getting_started/05_deploying_the_schema.md

---
title: Deploying the schema
summary: Deploy your GraphQL schema to a test or production environment
icon: rocket
---

# Getting started

[CHILDREN asList]

[alert]

> You are viewing docs for a pre-release version of silverstripe/graphql (4.x).

Reviewer: I'd drop the pre-release bit, it's very close to being stable

Reviewer 2: Removing that from all the docs globally is a separate issue. I think it pays to be consistent and have the warning on _all_ docs until we remove it from _all_ docs.

Help us improve it by joining #graphql on the [Community Slack](https://www.silverstripe.org/blog/community-slack-channel/),
and report any issues at [github.com/silverstripe/silverstripe-graphql](https://github.com/silverstripe/silverstripe-graphql).
Docs for the current stable version (3.x) can be found
[here](https://github.com/silverstripe/silverstripe-graphql/tree/3)
[/alert]

## Deploying the schema

One way or another, you must get the `.graphql-generated/` and `public/_graphql/` folders into your test and production environments for Silverstripe CMS (and your own custom queries) to work as expected. There are many ways to do so. The options below are listed in order of complexity.

### Options for any hosting solution

> #### Commit the schema to version control {#commit-to-vcs}

Reviewer: Putting this option first make it seem like it's the recommendation.  It's not, it's more like an option of last resort.  Better to start with 'on-demand' IMO since it's suitable for the majority of sites, and then offer other solution later for those that require the optimisation

Reviewer 2: > The options below are listed in order of complexity

This has been very clearly stated.

Reviewer: Do nothing and rely on the fallback is the least complex option so it should be first then?

Reviewer 2: That's in the "Multi-server hosting solutions" section, which makes it inherently more complex than the "Options for any hosting solution" approaches.

Reviewer: Talked offline - tl;dl

Add a section up the top saying:

"If your setup is a single webserver and run dev/build as part of your deployment pipeline, then you don't need to do anything. If you setup is something else, you'll need to implement one of these options."

Reviewer 2: Done. Also fixed a typo (graphsl should be graphql)

Be aware that for this approach to work, the process executing the `dev/build` must have write access to create the folders (or you must create those folders yourself, and give write access for those folders specifically), and for multi-server environments a `dev/build` or `dev/graphql/build` must be executed on each server hosting your site after each deployment.

#### Use a CI/CD pipeline to build your schema {#using-ci-cd}

Projects with more sophisticated requirements or bigger schemas exposing more than 100 `DataObject` classes may want to consider using a continuous-integration/continuous-deployment (CI/CD) pipeline to build their GraphQL schema.

In this kind of setup, you would need to update your deployment script to run the `dev/graphql/build` command which builds the `.graphql-generated/` and `public/_graphql/` folders. In multi-server environments this must be executed on each server hosting your site.

### Multi-server hosting solutions {#multi-server}

> If your site is hosted in an environment with multiple servers or configured to auto-scale with demand, there are some additional considerations. For example if you only generate the schema on one single server, then the other servers won’t have a `.graphql-generated/` or `public/_graphql/` folder (or those folders will be empty if you manually created them).

Reviewer: ```suggestion
If your site is hosted in an environment with multiple servers or configured to auto-scale with demand, there are some additional considerations. For example if you only generate the schema on one single server (typically via running `dev/build`), then the other servers won’t have a `.graphql-generated/` or `public/_graphql/` folder (or those folders will be empty if you manually created them).
```

Reviewer 2: done - though I mention `dev/graphql/build` here as well.

If you have a particularly large schema, you may want to ensure it is always built before the first GraphQL request. It might make sense for you to sync your `.graphql-generated/` and `public/_graphql/` folders across all your servers using an EFS or similar mechanism. In that case you only need to run `dev/build` or `dev/graphql/build` on the server with the original folder - but bear in mind that this may have a performance impact.

### Performance considerations when building the GraphQL schema {#performance-considerations}

The main driver in the resources it takes to build a GraphQL schema is the number DataObjects and the number of exposed relations in that schema. In most cases, not all DataObjects in your database will be included in your schema - best practice is to only add classes to your schema definition if you will need to query them. DataObjects not included in your schema will not impact the time or memory needed to build it.

Silverstripe CMS defines an "admin" schema it uses for its own purpose. This schema is relatively small and has a negligible performance impact.

> As an indication, we ran some benchmarks on a t3.micro AWS instance. The results are in the table below. These numbers may not be representative of the performance in your own environment. If you intend to build large GraphQL schemas, you should take the time to run your own benchmarks and adjust your deployment strategy accordingly.

Reviewer: ```suggestion
As an indication, benchmarks were run on a t3.micro AWS instance. The results are in the table below. These numbers may not be representative of the performance in your own environment. If you intend to build large GraphQL schemas, you should take the time to run your own benchmarks and adjust your deployment strategy accordingly.
```

Reviewer 2: done
